## Importing Libraries

In [5]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [7]:
path = r'C:\Users\Víctor A\OneDrive\DATA ANALITCS\EXCERCISES\IMMERSION\Instacart Basket Analysis'

In [8]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [10]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
1                        7                   15.0  
2                       12                   21.0  
3                        7                   29.0  
4                       15                   28.0  
...                    ...                    ...  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0  

[3421083 rows x 7 columns]

In [11]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

## What are Data Consistency Checks?

In [13]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## Mixed-Type Data

In [15]:
# Creating a new DataFrame to practice fixing mixed-type columns
df_test = pd.DataFrame()

In [16]:
# Creating a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [17]:
df_test

mix
0     a
1     b
2     1
3  True

In [18]:
# The function for checking whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Víctor A\AppData\Local\Temp\ipykernel_2224\2304849641.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [19]:
# Converting data type column to str
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

### Finding Missing Values

In [22]:
# Identifying and counting the number of null values (NaN which stands for 'Not a Number') in each column of a DataFrame

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
# Creating new dataframe, df_nan, containing only values within the "product_name" column that meet the condition isnull() = True

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values

In [26]:
# Imputing missing values with mean

# df['column with missings'].fillna(mean value, inplace=True)

In [27]:
# Imputing missing values with median

#df['column with missings'].fillna(median value, inplace=True)

In [28]:
# Since product_name datatype is string you can remove the missing values or filter out the ones that aren´t missing
# into a subset dataframe and continue your analysis with this new dataframe.


In [29]:
# First in order to compare the number of rows in your current dataframe

df_prods.shape

(49693, 5)

In [30]:
# Creating the new dataframe without the missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [31]:
# Checking the number of row in the new dataFrame

df_prods_clean.shape

(49677, 5)

In [32]:
# Another way you can drop all missing values is via the following command:

#df_prods.dropna(inplace = True)

# If you wanted to use this command to drop only the NaNs from a particular column, the code would look like this:

# df_prods.dropna(subset = [‘product_name’], inplace = True)

# In both cases, rather than creating an entirely new dataframe, you’re overwriting df_prods. 
# If inplace = False the command will only return a view of the changed dataframe, leaving the original dataframe untouched.

## Duplicates

### Finding Duplicates

In [35]:
# Creating a new dataFrame with only duplicates rows

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [36]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [38]:
# Checking the number of rows in the dataFrame

df_prods_clean.shape

(49677, 5)

In [39]:
# Creating a new dataframe that doesn’t include the duplicates you just identified using the drop_duplicates() function:

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [40]:
df_prods_clean_no_dups.shape

(49672, 5)

## Tidying Up and Exporting Changes

In [42]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task

### 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
    Tip: Keep an eye on min and max values!

In [45]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

#### - "Unnamed: 0"  column looks like it is an index and it is unnecessary for the analysis.
#### - The maximum value for "order_number" is 100, which seems high.
#### - The "days_since_last_order" has 206,209 missing values (difference between count of other columns and days_since_last_order column) which corresponds to the maximum "user_id". This indicates that there is a NaN value for each user's first order. We should either replace it with 0 or ask the client if they prefer to differentiate this data.
#### - "user_id" and "order_id" should be strings

### 3. Check for mixed-type data in your df_ords dataframe.

In [48]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### 5.  Run a check for missing values in your df_ords dataframe.

    In a markdown cell, report your findings and propose an explanation for any missing values you find

In [88]:
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

#### The "days_since_last_order" has 206209 missing values. Very likely this indicates that there is a NaN value for each user's first order.

### 6. Address the missing values using an appropriate method.

    In a markdown cell, explain why you used your method of choice

In [ ]:
# Create new variable 'is_first_order' that acts like a flag based on the missing value.

In [121]:
df_ords['is_first_order'] = df_ords['days_since_last_order'].isna()

In [125]:
# Check the results
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  is_first_order  
0                        8                    NaN            True  
1                        7                   15.0           False  
2                       12                   21.0           False  
3                        7                   29.0           False  
4                       15                   28.0           False  
...                    ...                    ...             ...  
3421078                 18                   29.0           False  
3421079                 10                   30.0           False  
3421080                 12                   18.0           False  
3421081                 12                    7.0           False  
3421082                 14                   30.0           False  

[3421083 rows x 8 columns]

In [133]:
# Create new subset where NaN values in 'days_since_last_order' are imputed with 0

df_ords_nonan = df_ords.copy()
df_ords_nonan['days_since_last_order'] = df_ords_nonan['days_since_last_order'].fillna(0, inplace = False)

In [135]:
df_ords_nonan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  is_first_order  
0                        8                    0.0            True  
1                        7                   15.0           False  
2                       12                   21.0           False  
3                        7                   29.0           False  
4                       15                   28.0           False  
...                    ...                    ...             ...  
3421078                 18                   29.0           False  
3421079                 10                   30.0           False  
3421080                 12                   18.0           False  
3421081                 12                    7.0           False  
3421082                 14                   30.0           False  

[3421083 rows x 8 columns]

####  I have decided to impute the missing values with 0 because, in my opinion, it increases the consistency of the data in the 'days_since_last_order' column and will aid in future analyses. If it becomes necessary in the future to differentiate between orders made on the same day and the first order, it will always be possible thanks to the 'order_number' variable. However, I have added the boolean variable 'is_first_order' to facilitate identification without having to apply the condition each time. Additionally, it makes the analysis more readable and self-explanatory.

### 7. Run a check for duplicate values in your df_ords data.

    In a markdown cell, report your findings and propose an explanation for any duplicate values you find

In [145]:
df_dups = df_ords_nonan[df_ords_nonan.duplicated()]

In [147]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, is_first_order]
Index: []

#### No duplicates were found.

### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [153]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [155]:
df_ords_nonan.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))